core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala

```diff
@@ -75,21 +95,31 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
   def acquireMemory(
       blockId: BlockId,
       numBytesToAcquire: Long,
-      numBytesToFree: Long): Boolean = lock.synchronized {
+      numBytesToFree: Long): Boolean = {
     assert(numBytesToAcquire >= 0)
     assert(numBytesToFree >= 0)
-    assert(memoryUsed <= poolSize)
-    if (numBytesToFree > 0) {
+    assert(lock.synchronized(memoryUsed <= poolSize))
```

Comment:
This lock acquisition should be cheap, so I think it's fine to have this assert.